### PR TITLE
introduce did framework

### PR DESCRIPTION
**Pull request by CoderZhi**
IDID.sol defines the interface and events

DIDStorage.sol stores dids and the corresponding uri and hash

AddressBasedDID.sol is for dids of did:io:...:${eth_address}

PebbleDID.sol is an example implementation of 3rd party did contract, which could only be updated by owner of the contract.

### CODE REVIEW

File: smartcontracts/contracts/DIDStorage.sol [hunk 1]
```diff
@@ -0,0 +1,35 @@
+pragma solidity >=0.4.22 <0.6;
```
Comment (lzxm160) [inline smartcontracts/contracts/DIDStorage.sol:1]: 0.6.0

File: smartcontracts/contracts/OwnableDID.sol [hunk 2]
```diff
@@ -0,0 +1,28 @@
+pragma solidity >=0.4.22 <0.6.0;
+
+import './DIDBase.sol';
+import './ownership/Ownable.sol';
+
+// OwnableDID is an implementation of IDID which is only updatable by owner
+contract OwnableDID is DIDBase, Ownable {
+    function verifyDID(string memory did) public view returns (bool);
+
+    function create(string memory did, bytes32 hash, string memory uri) public onlyOwner {
```
Comment (lzxm160) [inline smartcontracts/contracts/OwnableDID.sol:10]: do we need func like this createDIDSigned,so the device can sign a message and no need to call grpc?

File: smartcontracts/contracts/DIDStorage.sol [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+
+    function bumpNonce(address owner) public onlyOwner {
+        nonces[owner] = nonces[owner] + 1;
+    }
+
+    function upsert(string memory did, address owner, bytes32 hash, string memory uri) public onlyOwner {
+        dids[did] = DID(true, owner, hash, uri);
+    }
+
+    function deactivate(string memory did) public onlyOwner {
```
Comment (raullenchai) [inline smartcontracts/contracts/DIDStorage.sol:32]: then we need function activate()?

can we call upsert with false instead?

File: smartcontracts/contracts/DIDStorage.sol [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+    function getNonce(address owner) public view returns (uint256) {
+        return nonces[owner];
+    }
+
+    function bumpNonce(address owner) public onlyOwner {
+        nonces[owner] = nonces[owner] + 1;
+    }
+
+    function upsert(string memory did, address owner, bytes32 hash, string memory uri) public onlyOwner {
+        dids[did] = DID(true, owner, hash, uri);
```
Comment (raullenchai) [inline smartcontracts/contracts/DIDStorage.sol:29]: len(uri) shall be capped?

File: smartcontracts/contracts/DIDBase.sol [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+            db = DIDStorage(dbAddr);
+        }
+    }
+
+    function transferDBOwnership(address newOwner) public onlyOwner {
+        db.transferOwnership(newOwner);
+    }
+
+    function getHash(string memory did) public view returns (bytes32) {
+        require(db.exist(did), "did does not exist");
```
Comment (raullenchai) [inline smartcontracts/contracts/DIDBase.sol:25]: nit: DID

same below

**Pull request merged**